### ISSUE TITLE
hail iallocator: disk_space_total:JSNull: Unable to read Int

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#271]
Originally reported of Google Code with ID 215.

```
What software version are you running? Please provide the output of "gnt-
cluster --version" and "gnt-cluster version".

multiverse ~ # gnt-cluster --version
gnt-cluster (ganeti v2.5.0rc5) 2.5.0~rc5
multiverse ~ # gnt-cluster version
Software version: 2.5.0~rc5
Internode protocol: 2050000
Configuration format: 2050000
OS api version: 20
Export interface: 0


<b>What distribution are you using?</b>

Sabayon (gentoo), with package compiled and built using https://github.com/optiz0r/gentoo-overlay/blob/master/app-emulation/ganeti/ganeti-2.5.0_rc5.ebuild


<b>What steps will reproduce the problem?</b>

Attempt to create a new virtual machine using the hail iallocator and file-based disks, or execute the following debug command (substituting the os-type for a suitable local value):

gnt-debug iallocator --debug --dir=out -I hail --mode=allocate --memory=512 --disks=8192 --disk-template=file --vcpus=1 --os-type=image+sabayon-kvm-7 ganeti-test.fqdn


<b>What is the expected output? What do you see instead?</b>

Expect hail iallocator to correctly identify a destination node and for install to progress. Actual output is the following error:

OpExecError: ("Failure while running the iallocator script: iallocator
module 'hail' failed: exited with exit code 1, output 'Error: invalid
data for instance 'ganeti-test.fqdn': key 'disk_space_total', value
'JSNull': Unable to read Int\n'",)
Failure: command execution error:
("Failure while running the iallocator script: iallocator module
'hail' failed: exited with exit code 1, output 'Error: invalid data
for instance 'ganeti-test.fqdn': key 'disk_space_total', value
'JSNull': Unable to read Int\n'",)


<b>Please provide any additional information below.</b>

Error is in the _ComputeDiskSize (and also _ComputeDiskSizePerVG) functions of cmdlib.py. Attached patch against 2.5.0rc5 version of cmdlib.py will resolve the issue (has not been corrected in git HEAD).

I had also reported this error via the mailing list, but as I now have the patch, thought it would be better to repost here.
```
Originally added on 2012-02-10 19:53:42 +0000 UTC.

Attachments: 
[fix-file-disk-size.patch](https://storage.googleapis.com/google-code-attachments/ganeti/issue-215/comment-0/fix-file-disk-size.patch)

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi,

I'm not sure the patch works correctly (the issue is a bit more complex).

Since gnt-node list won't report free disk space for file storage (or shared file), this will allow hail to complete, but it will not have real data to work with. I expect that on a cluster which doesn't have at all LVM, just file storage, hail will still be broken, as the node won't report anything in `total_disk`.

I've started looking at making htools work better with non-LVM based storage, but it will require more changes to the iallocator protocol than just this, I think.
```
Originally added on 2012-02-20 10:05:37 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi Iustin,

I know the storage reporting is still lacking, but was this issue fixed in the meantime. If so, we can close it, if not we can assign it to Helga, as she'll be working on the storage reporting.

Thanks,
Guido

```
Originally added on 2013-02-13 14:44:04 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
As far as I know, the issue is still present. I'll test further.
```
Originally added on 2013-02-13 14:52:21 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Helga, could you confirm whether this now works, and mark the bug as fixed, in case?
I believe now we finally have lvm as optional, correct?

Thanks,

Guido

```
Originally added on 2013-12-09 14:00:39 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I checked and I think this is fixed in 2.9. Note that it is necessary to set the list of cluster enabled disk templates so that 'file' is the default template. This can be done by setting 'file' as first template:

gnt-cluster modify --enabled-disk-templates=file[,other templates]

(You might need to adjust the --ipolicy-disk-templates as well.)

It is necessary to set the file template as default template, because the iallocator can only handle one disk template and by setting the default you tell it to use file storage. If we wanted more flexibility, we would need to make bigger changes in the iallocator protocol as iustin mentioned.
```
Originally added on 2013-12-11 08:55:24 +0000 UTC.

Changed State: Fixed
